service-userlayer/src/main/java/org/oskari/map/userlayer/input/FeatureCollectionParsers.java

```diff
@@ -0,0 +1,86 @@
+package org.oskari.map.userlayer.input;
+
+import org.geotools.data.simple.SimpleFeatureCollection;
+import org.geotools.data.simple.SimpleFeatureIterator;
+import org.geotools.data.simple.SimpleFeatureSource;
+import org.geotools.feature.DefaultFeatureCollection;
+import org.geotools.geometry.jts.JTS;
+import org.geotools.referencing.CRS;
+import org.opengis.feature.simple.SimpleFeature;
+import org.opengis.referencing.FactoryException;
+import org.opengis.referencing.crs.CoordinateReferenceSystem;
+import org.opengis.referencing.operation.MathTransform;
+
+import com.vividsolutions.jts.geom.Geometry;
+
+public class FeatureCollectionParsers {
+
+    private FeatureCollectionParsers() {}
+
+    public static boolean hasByFileExt(String fileExt) {
+        if (fileExt == null) {
+            return false;
+        }
+        fileExt = fileExt.toUpperCase();
+        switch (fileExt) {
+        case GPXParser.SUFFIX:
+        case KMLParser.SUFFIX:
+        case MIFParser.SUFFIX:
+        case SHPParser.SUFFIX:
+            return true;
+        default:
+            return false;
+        }
+    }
+
+    public static FeatureCollectionParser getByFileExt(String fileExt) {
+        if (fileExt == null) {
+            return null;
+        }
+        fileExt = fileExt.toUpperCase();
+        switch (fileExt) {
+        case GPXParser.SUFFIX: return new GPXParser();
+        case KMLParser.SUFFIX: return new KMLParser();
+        case MIFParser.SUFFIX: return new MIFParser();
+        case SHPParser.SUFFIX: return new SHPParser();
+        default: return null;
+        }
+    }
+
+    /**
+     * Read Features from FeatureSource to memory (ensure they won't
+     * disappear when DataStore#dispose() is called) while transforming
+     * their geometries from source projection to target projection
+     * @throws Exception lots can go wrong
+     */
+    public static SimpleFeatureCollection read(SimpleFeatureSource src,
+            CoordinateReferenceSystem sourceCRS, CoordinateReferenceSystem targetCRS) throws Exception {
+        DefaultFeatureCollection fc = new DefaultFeatureCollection();
+        SimpleFeatureCollection sfc = src.getFeatures();
+        MathTransform transform = getTransform(sourceCRS, targetCRS);
+        try (SimpleFeatureIterator it = sfc.features()) {
+            while (it.hasNext()) {
+                SimpleFeature f = it.next();
+                if (transform != null) {
```

Comment:
You have the null check here since you want to read the features even if transform isn't done? I think not doing the transform should be handled in some other way than just passing features through. Not sure if not having default geometry is handled somewhere down the line, but that's kind of a deal breaker as well.

Reply:
Having no geometry is currently checked and tracked at the `UserLayerDataService` level. The number of features that get declined because they had no geometry is currently reported back to the client. However I'm not certain how valuable that information truly is.

Reply:
Agree with the transform check, we must always know both the sourceCRS and the targetCRS, otherwise none of this makes any sense.